elasticdl/python/master/master.py

```diff
@@ -0,0 +1,401 @@
+        self.model_module = load_module(
+            get_module_file_path(args.model_zoo, args.model_def)
+        ).__dict__
+        self.model_inst = load_model_from_module(
+            args.model_def, self.model_module, args.model_params
+        )
+        self.optimizer = self.model_module[args.optimizer]()
+
+        # Initialize checkpoint service
+        self.checkpoint_service = self._create_checkpoint_service(args)
+
+        # Initialize evaluation service
+        self.evaluation_service = self._create_evaluation_service(args)
+
+        # Initialize embedding service
+        (
+            self.embedding_service_endpoint,
+            self.embedding_dims,
+        ) = self._create_embedding_service(args)
+
+        # Initialize master service
+        self.master_servicer, self.server = self._create_master_service(args)
+
+        # Initialize worker manager
+        self.worker_manager = self._create_worker_manager(args)
+
+    def prepare(self):
+        '''
+        Start the components one by one.
+        Make sure that the entire cluster is ready to run.
+        '''
+        # Start the evaluation service if requested
+        if self.evaluation_service:
+            self.logger.info("Starting evaluation service")
+            self.evaluation_service.start()
+            self.logger.info("Evaluation service started")
+
+        # Start the master GRPC server
+        self.logger.info("Starting master RPC server")
+        self.server.start()
+        self.logger.info("Master RPC server started")
+
+        # Start the worker manager if requested
+        if self.worker_manager:
+            self.worker_manager.update_status(WorkerManagerStatus.PENDING)
+            self.worker_manager.start_workers()
+            self.worker_manager.update_status(WorkerManagerStatus.RUNNING)
+
+        # Start TensorBoard k8s Service if requested
+        if self.tb_service and self.tb_client:
+            self.logger.info("Starting tensorboard service")
+            self.tb_service.start()
+            self.tb_client.start_tensorboard_service()
+            self.logger.info("Tensorboard service started")
+
+    def run(self):
+        '''
+        The main loop of master.
+        Dispatch the tasks to the workers until all the tasks are completed.
+        '''
+        try:
+            while True:
+                if self.task_d.finished():
+                    if self.worker_manager:
+                        self.worker_manager.update_status(
+                            WorkerManagerStatus.FINISHED
+                        )
+                    if self.checkpoint_output_path:
+                        self.master_servicer.save_latest_checkpoint(
+                            self.checkpoint_output_path
+                        )
+                    break
+                time.sleep(30)
+        except KeyboardInterrupt:
+            self.logger.warning("Server stopping")
+
+        self._stop()
```

Comment:
This isn't guaranteed to be called. Perhaps move to a `finally` block.

Reply:
Good catch, updated.